### PR TITLE
[PyTorch] [gradcheck] change backward() to grad()

### PR DESCRIPTION
**Pull request by ssnl**
Change backward calls to grad to avoid memory leak from #7343 
Replace unnecessary create_graph=True with retain_graph=True

The memory leak is blocking #7270 .

### CODE REVIEW

**Comment by SsnL** [comment]
The current gradgradcheck is broken because the inputs, containing `grad_outputs`, is not actually used in the function. Then when it is non-contiguous and gradcheck makes it contiguous, things break. 
```py
>>> # print numerical and analytical grad wrt the grad_output
>>> torch.autograd.gradgradcheck(lambda x: x, [torch.randn(3).requires_grad_()], gen_non_contig_grad_outputs=False)
(tensor([[ 1.0133,  0.0000,  0.0000],
        [ 0.0000,  1.0133,  0.0000],
        [ 0.0000,  0.0000,  1.0133]]), tensor([[ 1.,  0.,  0.],
        [ 0.,  1.,  0.],
        [ 0.,  0.,  1.]]))

>>> torch.autograd.gradgradcheck(lambda x: x, [torch.randn(3).requires_grad_()], gen_non_contig_grad_outputs=True)
(tensor([[ 0.,  0.,  0.],
        [ 0.,  0.,  0.],
        [ 0.,  0.,  0.]]), tensor([[ 0.,  0.,  0.],
        [ 0.,  0.,  0.],
        [ 0.,  0.,  0.]]))
```

Also fixing this.

File: torch/autograd/gradcheck.py [hunk 1]
```diff
@@ -59,10 +49,13 @@ def contiguous(input):
...
 
 
+# `input` is input to `fn`
+# `target` is the Tensors wrt whom Jacobians are calculated
 def get_numerical_jacobian(fn, input, target, eps=1e-3):
     # To be able to use .view(-1) input must be contiguous
     input = contiguous(input)
-    target = contiguous(target)
+    if not all(t.is_contiguous() for t in iter_tensors(target)):
+        raise RuntimeError("target must only contain contiguous Tensors")
```
Comment (apaszke) [inline torch/autograd/gradcheck.py]: Why was `contiguous` invalid in here? It computes the derivative numerically so this shouldn't really matter

File: torch/autograd/gradcheck.py [hunk 2]
```diff
@@ -267,7 +265,8 @@ def new_func(*input_args):
         input_args = input_args[:-len(grad_outputs)]
         outputs = _differentiable_outputs(func(*input_args))
         input_args = tuple(x for x in input_args if isinstance(x, torch.Tensor) and x.requires_grad)
-        grad_inputs = torch.autograd.grad(outputs, input_args, grad_outputs, create_graph=True)
+        grad_inputs = torch.autograd.grad(outputs, input_args, grad_outputs,
+                                          create_graph=True, only_inputs=True, allow_unused=True)
```
Comment (apaszke) [inline torch/autograd/gradcheck.py]: This got an extra `allow_unused` flag - why? Also, `only_inputs` is `True` by default IIRC, so no need to specify it.

File: torch/autograd/gradcheck.py [hunk 3]
```diff
@@ -200,18 +197,19 @@ def fn(input):
     zero_gradients(inputs)
```
Comment (apaszke) [inline torch/autograd/gradcheck.py]: This can be removed now I think

File: torch/autograd/gradcheck.py (hunk 1 above)
Comment (SsnL) [reply torch/autograd/gradcheck.py]: The version here is WIP. But when `target` is not `input`, e.g., in some tests we have `get_numerical_jacobian(module, input, module_params)`, then making them contiguous is a copy, and changing values in it has no effect on the resulting gradients, and thus the numerical result would be 0.

The current gradgradcheck is broken because of this. It only passed because we had the same `contiguous(inputs)` in calculating analytical grads.

**Comment by SsnL** [comment]
I just got rid of the contiguity inforcement altogether. Hope that everything still works.

File: tools/autograd/templates/Functions.cpp [hunk 4]
```diff
@@ -852,7 +852,7 @@ std::tuple<Tensor, Tensor, Tensor> prelu_double_backward(
       return std::tuple<Tensor, Tensor, Tensor>(
                 ggO,
                 ggW_maybe_squeeze.expand_as(gO) * gO * nonpositive_mask,
-                (ggI * gO * nonpositive_mask).sum()
+                (ggI * gO * nonpositive_mask).sum().expand_as(weight)
```
Comment (SsnL) [inline tools/autograd/templates/Functions.cpp]: Fixing a bug in PReLU double backward here

**Comment by SsnL** [comment]
@apaszke This is ready for review.

File: tools/autograd/templates/Functions.cpp (hunk 4 above)
Comment (apaszke) [reply tools/autograd/templates/Functions.cpp]: Nice!

File: torch/autograd/gradcheck.py [hunk 5]
```diff
@@ -74,23 +61,25 @@ def get_numerical_jacobian(fn, input, target, eps=1e-3):
...
     # TODO: compare structure
     for x_tensor, d_tensor in zip(x_tensors, j_tensors):
-        flat_tensor = x_tensor.view(-1).detach()
-        for i in range(flat_tensor.nelement()):
-            orig = flat_tensor[i].item()
-            flat_tensor[i] = orig - eps
+        # need data here to get around the version check because without .data,
+        # the following code updates version but doesn't change content
+        x_tensor = x_tensor.data
+        for d_idx, x_idx in enumerate(product(*[range(m) for m in x_tensor.size()])):
```
Comment (apaszke) [inline torch/autograd/gradcheck.py]: Nice workaround!

File: torch/autograd/gradcheck.py [hunk 6]
```diff
@@ -254,17 +243,19 @@ def gradgradcheck(func, inputs, grad_outputs=None, eps=1e-6, atol=1e-5, rtol=1e-
...
         # shape, type, and device as the outputs
         def randn_like(x):
-            var = torch.testing.randn_like(x if x.is_floating_point() else x.double())
+            y = torch.testing.randn_like(x if x.is_floating_point() else x.double(), requires_grad=True)
             if gen_non_contig_grad_outputs:
-                var = torch.testing.make_non_contiguous(var)
-            var.requires_grad = True
-            return var
+                y = torch.testing.make_non_contiguous(y).requires_grad_()
+            return y
```
Comment (apaszke) [inline torch/autograd/gradcheck.py]: nit: you could remove `requires_grad` from all parts before this line and just do `return y.requires_grad_()`

File: torch/autograd/gradcheck.py [hunk 7]
```diff
@@ -254,17 +243,19 @@ def gradgradcheck(func, inputs, grad_outputs=None, eps=1e-6, atol=1e-5, rtol=1e-
...
         grad_outputs_gen = (randn_like(x) for x in outputs)
         grad_outputs = list(grad_outputs_gen) if not isinstance(inputs, tuple) else tuple(grad_outputs_gen)
 
-    def new_func(*input_args):
-        input_args = input_args[:-len(grad_outputs)]
+    num_outputs = len(grad_outputs)
+
+    def new_func(*args):
+        input_args = args[:-num_outputs]
+        grad_outputs = args[-num_outputs:]
```
Comment (apaszke) [inline torch/autograd/gradcheck.py]: Why didn't we need that previously? Do they change at different runs?

**Review by apaszke — APPROVED** [review]
LGTM. One concern about `gradgradcheck` that would be good to clarify before merging.

**Review by apaszke — APPROVED** [review]
